modules/mono/glue/GodotSharp/GodotSharp/Core/Callable.generics.cs

```diff
@@ -0,0 +1,489 @@
+using System;
+using Godot.NativeInterop;
+
+namespace Godot;
+
+public readonly partial struct Callable
+{
+    internal static ArgumentException InvalidArgCountException(string paramName, int countExpected, int countReceived)
+    {
+        return new ArgumentException(
+            "Invalid argument count for invoking callable." +
+            $" Expected {countExpected} arguments, received {countReceived}.",
+            paramName);
+    }
```

Comment:
Since you are making this a method, how about making the condition check part of the method as well?

```csharp
internal static void ThrowIfArgCountMismatch(int countExpected, NativeVariantPtrArgs args, [CallerArgumentExpression("args")] string? paramName = null)
{
    if (args.Count != countExpected)
    {
        throw new ArgumentException(
            "Invalid argument count for invoking callable." +
            $" Expected {countExpected} arguments, received {args.Count}.",
            paramName);
    }
}
```

Then it can be used like this:

```csharp
Callable.ThrowIfArgCountMismatch(1, args);
```

Reply:
Sounds good, although I'll still make the method return the exception instead of taking care of throwing. It's what I've seen in the .NET libraries, and it looks more readable to me. It may be better for code flow analysis as well, unless the compiler can understand that the throwing method never returns.

Reply:
If the method always throws you can use the `[DoesNotReturn]` attribute, but if we include the condition in the method then it doesn't always throw, so you have to assume that the method may return.

> It's what I've seen in the .NET libraries.

That's interesting, I made the suggestion thinking about [`ArgumentException.ThrowIfNullOrEmpty`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/ArgumentException.cs,108). This pattern allows to inline methods that call it[^1] although that doesn't apply in our case.

[^1]: Methods that throw are not inlined so by moving the throwing to a different method (e.g.: `ThrowIfNullOrEmpty`) the original method can be inlined.

Reply:
I didn't know there were public methods like that. The ones I saw are internal: https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Error.cs

I suppose it's not that consistent. In any case, I forgot the condition is now inside the method, so the argument about reachability doesn't make much sense. Now what worries me is the overhead, if for some reason the compiler doesn't inline the call, although it should be smart enough to (specially if they have public APIs like that).

Reply:
I checked the JIT output and it inlines the call just fine. I expect the AOT compiler to do so as well. Interestingly, the `ThrowIf` method results in less instructions as it does the actual throwing in another method that's not inlined.